File: backend/SurveyViewerService/SurveyViewerService/Logic/PubSub.cs
namespace SurveyViewerService.Logic
{
	using System;
	using System.Linq;
	using System.Threading.Tasks;
	using Google.Cloud.PubSub.V1;
	using Newtonsoft.Json;
	using SurveyViewerService.Contracts;
	using SurveyViewerService.Model.SaveSurveyResult;

	/// <summary>
	///   Access google cloud Pub/Sub.
	/// </summary>
	public class PubSub : IPubSub
	{
		private readonly IConfiguration configuration;

		/// <summary>
		///   Access google cloud Pub/Sub.
		/// </summary>
		private PublisherClient client;

		/// <summary>
		///   Creates a new instance of <see cref="PubSub" />.
		/// </summary>
		/// <param name="configuration">Access the application configuration.</param>
		public PubSub(IConfiguration configuration)
		{
			this.configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
		}

		/// <summary>
		///   Publish a message to Pub/Sub to save the survey result.
		/// </summary>
		/// <param name="surveyId">The id of the survey.</param>
		/// <param name="surveySubmitResult">The result of the survey.</param>
		/// <returns>A <see cref="Task" />.</returns>
		public async Task PublishMessageAsync(string surveyId, ISurveySubmitResult surveySubmitResult)
		{
			if (this.client == null)
			{
				var topic = TopicName.FromProjectTopic(this.configuration.ProjectId, this.configuration.PubSubTopic);
				this.client = await PublisherClient.CreateAsync(topic);
			}

			var request = new SaveSurveyResultRequest(
				surveySubmitResult.Questions.Select(
					q => new SurveyResult
					{
						QuestionId = q.QuestionId,
						QuestionValue = int.Parse(q.Value)
					}))
			{
				ParticipantId = surveySubmitResult.ParticipantId,
				SurveyId = surveyId
			};

			var json = JsonConvert.SerializeObject(request);

			await this.client.PublishAsync(json);
		}
	}
}

File: frontend/SurveyVote/application/services/pub-sub.js
const { PubSub } = require('@google-cloud/pubsub');
const uuid = require('uuid');

const pubSubClient = new PubSub();

/**
 * Initializes a pub/sub client.
 * @param {object} options An options object.
 * @param {string} options.topicName The name of the pub/sub topic.
 * @returns A new pub/sub client.
 */
const initialize = (options) => {
  const { topicName } = options;

  const client = {
    /**
     * Publish a message to pub/sub.
     * @param {object} message The message that is sent.
     * @param {string} message.internalSurveyId The internal survey id.
     * @param {string} message.participantId The id of the survey participant.
     * @param {Array} message.results The survey result of the participant.
     *  Entries are objects of { questionId, choiceId }.
     */
    publish: async (message) => {
      const {
        internalSurveyId,
        participantId,
        results,
      } = message;

      const resultsJson = Array.from(results.map(({ questionId, choiceId }) => `{"questionId":"${questionId}","choiceId":"${choiceId}"}`)).join(',');
      const json = `{"processId":"${uuid.v4()}","surveyResult":{"documentId":"","created":"","parentDocumentId":"${internalSurveyId}","participantId":"${participantId}","isSuggested":false,"results":[${resultsJson}]}}`;
      const data = Buffer.from(json);
      await pubSubClient
        .topic(topicName)
        .publishMessage({ data });
    },
  };

  return client;
};

module.exports = initialize;
